xarray/core/dataset.py

```diff
@@ -1433,11 +1449,68 @@ def __setitem__(self, key: Hashable, value) -> None:
         variable.
         """
         if utils.is_dict_like(key):
-            raise NotImplementedError(
-                "cannot yet use a dictionary as a key to set Dataset values"
-            )
+            from .dataarray import DataArray
 
-        self.update({key: value})
+            # check for consistency first
+            if isinstance(value, Dataset):
+                missing_vars = [
+                    str(name) for name in value.data_vars if name not in self.data_vars
+                ]
+                if len(missing_vars) > 0:
+                    raise KeyError(
+                        "Variables {} in new values not available in dataset!".format(
+                            ",".join(missing_vars)
+                        )
+                    )
+            for name, var in self.items():
+                missing_keys = [k for k in key.keys() if k not in var.dims]
+                if len(missing_keys) > 0:
+                    raise KeyError(
+                        "Variable {} does not contain dimensions {}!".format(
+                            name, ",".join(missing_keys)
+                        )
+                    )
+
+                # test indexing
+                try:
+                    var_k = var[key]
+                except IndexError:
+                    raise IndexError(
+                        "Indexer {} not available in variable '{}'".format(key, name)
+                    )
+
+                if isinstance(value, Dataset):
+                    val = value[name]
+                else:
+                    val = value
+                if isinstance(val, DataArray):
+                    for dim in val.dims:
+                        coord1 = var_k[dim]
+                        coord2 = val[dim]
+                        if (len(coord1) != len(coord2)) or any(
+                            coord1.values != coord2.values
+                        ):
+                            raise IndexError(
+                                "Variable '{}': dimension coordinate '{}' conflicts"
+                                " between indexed and indexing objects: "
+                                " {}\n vs.\n {}".format(name, dim, coord2, coord1)
+                            )
+                elif isinstance(val, np.ndarray):
+                    try:
+                        np.broadcast_to(val, var_k.shape)
+                    except ValueError:
+                        raise ValueError(
+                            "Variable '{}': input array of shape {} cannot be broadcast"
+                            " to shape {}".format(name, val.shape, var_k.shape)
+                        )
```

Comment:
Generally cases where we check for "broadcasting shape" rather than "matching dimensions" is an example of "code smell" in Xarray. There are a few places where we do it in the existing codebase, but I would prefer not to add more. One of the main selling points of Xarray is that NumPy's broadcasting rules are hard to understand and use correctly.

So I'm not sure we need to support assigning a NumPy array directly. We could potentially raise an error for assigning anything other than a DataArray or a scalar value, which would be simpler and less error prone.

Reply:
ok, I deleted the numpy array functionality and included an error instead